activation/activation.go

```diff
@@ -911,54 +899,44 @@ func (b *Builder) broadcast(ctx context.Context, atx scale.Encodable) (int, erro
 	return len(buf), nil
 }
 
-// searchPositioningAtx returns atx id with the highest tick height.
-// publish epoch is used for caching the positioning atx.
-func (b *Builder) searchPositioningAtx(
+// find returns atx id with the highest tick height.
+// The publish epoch (of the built ATX) is used for:
+// - caching the positioning atx,
+// - filtering candidates for positioning atx (it must be published in an earlier epoch than built ATX).
+//
+// It always returns an ATX, falling back to the golden one as the last resort.
+func (f *positioningAtxFinder) find(
 	ctx context.Context,
-	nodeID types.NodeID,
+	atxs AtxService,
 	publish types.EpochID,
-) (types.ATXID, error) {
-	logger := b.logger.With(log.ZShortStringer("smesherID", nodeID), zap.Uint32("publish epoch", publish.Uint32()))
+) types.ATXID {
+	logger := f.logger.With(zap.Uint32("publish epoch", publish.Uint32()))
 
-	b.posAtxFinder.finding.Lock()
-	defer b.posAtxFinder.finding.Unlock()
+	f.finding.Lock()
+	defer f.finding.Unlock()
 
-	if found := b.posAtxFinder.found; found != nil && found.forPublish == publish {
+	if found := f.found; found != nil && found.forPublish == publish {
 		logger.Debug("using cached positioning atx", log.ZShortStringer("atx_id", found.id))
-		return found.id, nil
+		return found.id
 	}
 
-	latestPublished, err := atxs.LatestEpoch(b.db)
-	if err != nil {
-		return types.EmptyATXID, fmt.Errorf("get latest epoch: %w", err)
-	}
-
-	logger.Info("searching for positioning atx", zap.Uint32("latest_epoch", latestPublished.Uint32()))
-
-	// positioning ATX publish epoch must be lower than the publish epoch of built ATX
-	positioningAtxPublished := min(latestPublished, publish-1)
-	id, err := findFullyValidHighTickAtx(
-		ctx,
-		b.atxsdata,
-		positioningAtxPublished,
-		b.conf.GoldenATXID,
-		b.validator,
-		logger,
-		VerifyChainOpts.AssumeValidBefore(time.Now().Add(-b.postValidityDelay)),
-		VerifyChainOpts.WithTrustedID(nodeID),
-		VerifyChainOpts.WithLogger(b.logger),
-	)
+	id, err := atxs.PositioningATX(ctx, publish-1)
 	if err != nil {
-		logger.Info("search failed - using golden atx as positioning atx", zap.Error(err))
-		id = b.conf.GoldenATXID
+		logger.Warn("failed to get positioning ATX - falling back to golden", zap.Error(err))
+		f.found = &struct {
+			id         types.ATXID
+			forPublish types.EpochID
+		}{f.golden, publish}
```

Comment:
Given that this struct definition is written out more than once, maybe it deserves a named type?